src/celeritas/optical/ScintillationParams.cc

```diff
@@ -36,11 +39,67 @@ ScintillationParams::from_import(ImportData const& data)
         return nullptr;
     }
 
+    auto const& num_optmats = data.optical.size();
+
     Input input;
-    for (auto const& mat : data.optical)
+    input.resolution_scale.resize(num_optmats);
+    if (data.optical_params.scintillation_by_particle)
+    {
+        // Collect ScintillationParticleIds
+        input.pid_to_scintpid.resize(data.particles.size());
+        ScintillationParticleId scintpid;
+        for (auto const& [matid, iom] : data.optical)
+        {
+            auto const& iomsp = iom.scintillation.particles;
+            for (auto const& [pdg, ipss] : iomsp)
+            {
+                auto const pid = particle_params->find(PDGNumber{pdg});
+                // Add new ScintillationParticleId
+                input.pid_to_scintpid[pid.get()] = scintpid++;
+            }
+        }
+        // Resize particle- and material-dependent spectra
+        input.particles.resize(input.pid_to_scintpid.size() * num_optmats);
+    }
+    else
     {
-        input.data.push_back(mat.second.scintillation);
+        // Resize material-only spectra
+        input.materials.resize(num_optmats);
     }
+
+    size_type optmatidx{0};
+    size_type scintpidx{0};
+    for (auto const& [matid, iom] : data.optical)
+    {
+        input.resolution_scale[optmatidx] = iom.scintillation.resolution_scale;
+
+        if (!data.optical_params.scintillation_by_particle)
+        {
+            // Material spectrum
+            auto const& iomsm = iom.scintillation.material;
+            ImportMaterialScintSpectrum mat_spec;
+            mat_spec.yield_per_energy = iomsm.yield_per_energy;
+            mat_spec.components = iomsm.components;
+            input.materials[optmatidx] = std::move(mat_spec);
+        }
+        else
+        {
+            // Particle and material spectrum
+            auto const& iomsp = iom.scintillation.particles;
+
+            for (auto const& [pdg, ipss] : iomsp)
+            {
+                ImportParticleScintSpectrum part_spec;
+                part_spec.yield_vector = ipss.yield_vector;
+                part_spec.components = ipss.components;
+                input.particles[num_optmats * scintpidx + optmatidx]
+                    = std::move(part_spec);
+                scintpidx++;
+            }
```

Comment:
```suggestion
            for (auto const& [pdg, ipss] : iomsp)
            {
                if (auto const pid = particle_params->find(PDGNumber{pdg}))
                {   
                    auto scintpid = input.pid_to_scintpid[pid.get()];
                    CELER_ASSERT(scintpid);
                    ImportParticleScintSpectrum part_spec;
                    part_spec.yield_vector = ipss.yield_vector;
                    part_spec.components = ipss.components;
                    input.particles[num_optmats * scintpid.get() + optmatidx]
                        = std::move(part_spec);
                } 
            }
```
`input.particles` won't necessarily have the same size as `iom.scintillation.particles` (which can have up to the number of particles defined in `G4MaterialPropertiesIndex`, most of which aren't currently supported in Celeritas).